process/rating/peerHonesty/peerScore.go

```diff
@@ -0,0 +1,280 @@
+			process.ErrInvalidMinScore,
+		)
+	}
+	if peerHonestyConfig.MaxScore < 0 {
+		return fmt.Errorf("%w, MaxScore value should be positive or zero",
+			process.ErrInvalidMaxScore,
+		)
+	}
+	if peerHonestyConfig.UnitValue < 0 {
+		return fmt.Errorf("%w, UnitValue value should be positive or zero",
+			process.ErrInvalidUnitValue,
+		)
+	}
+
+	isBadPeerThresholdOk := peerHonestyConfig.BadPeerThreshold < 0 && peerHonestyConfig.MinScore < peerHonestyConfig.BadPeerThreshold
+	if !isBadPeerThresholdOk {
+		return fmt.Errorf("%w, BadPeerThreshold value should be in interval (MinScore, 0)",
+			process.ErrInvalidBadPeerThreshold,
+		)
+	}
+
+	return nil
+}
+
+func (pph *p2pPeerHonesty) executeDecay(ctx context.Context, handler func()) {
+	for {
+		select {
+		case <-time.After(pph.updateIntervalForDecay):
+			handler()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (pph *p2pPeerHonesty) applyDecay() {
+	pph.mut.Lock()
+	defer pph.mut.Unlock()
+
+	keys := pph.cache.Keys()
+	for _, key := range keys {
+		psObj, _ := pph.cache.Get(key)
+		if psObj == nil {
+			continue
+		}
+
+		ps, ok := psObj.(*peerScore)
+		if !ok {
+			continue
+		}
+
+		for topic, score := range ps.scores {
+			score = score * pph.decayCoefficient
+			if score < approximateZero && score > 0 {
+				score = 0
+			}
+
+			ps.scores[topic] = score
+		}
+	}
+}
+
+// ChangeScore will change the score of a public key on a provided topic
+func (pph *p2pPeerHonesty) ChangeScore(pk string, topic string, units int) {
+	pph.mut.Lock()
+	defer pph.mut.Unlock()
+
+	ps := pph.getValidPeerScore(pk)
+
+	oldValue := ps.scores[topic]
+	change := float64(units) * pph.unitValue
+
+	if change < 0 {
+		//TODO switch this to log.Trace in the future
+		log.Debug("p2pPeerHonesty.ChangeScore decrease",
+			"pk", hex.EncodeToString([]byte(pk)),
+			"current", fmt.Sprintf("%.2f", oldValue),
+			"change", fmt.Sprintf("%.2f", change),
+		)
+	}
+
+	ps.scores[topic] += change
+
+	pph.checkBlacklist(ps)
+}
+
+func (pph *p2pPeerHonesty) getValidPeerScore(pk string) *peerScore {
+	key := []byte(pk)
+
+	var ps *peerScore
+	psObj, _ := pph.cache.Get(key)
+	if psObj == nil {
+		ps = pph.createDefaultPeerScore(pk)
+		return ps
+	}
+
+	ps, ok := psObj.(*peerScore)
+	if !ok {
+		ps = pph.createDefaultPeerScore(pk)
+	}
+
+	return ps
+}
+
+func (pph *p2pPeerHonesty) createDefaultPeerScore(pk string) *peerScore {
+	key := []byte(pk)
+
+	ps := &peerScore{
+		scores: make(map[string]float64),
+		pk:     pk,
+	}
+	pph.cache.Put(key, ps, ps.size())
+
+	return ps
+}
+
+func (pph *p2pPeerHonesty) checkBlacklist(ps *peerScore) {
+	shouldBlacklist := false
+	for topic, score := range ps.scores {
+		if score > pph.maxScore {
+			ps.scores[topic] = pph.maxScore
+			continue
+		}
+
+		if score < pph.minScore {
+			ps.scores[topic] = pph.minScore
+		}
+
+		if score < pph.badPeerThreshold {
+			shouldBlacklist = true
+		}
+	}
+
+	if !shouldBlacklist {
+		return
+	}
+
+	if pph.blackListedPkCache.Has(ps.pk) {
```

Comment:
So we do not extend the `stay in the black list` period for already-blacklisted peers even if `shouldBlacklist == true`, right? Well, it makes sense since a blacklisted peer can't misbehave anymore anyway - at least not with effect (I am just thinking out loud).

Reply:
exactly, since this is checked on each 10 seconds or so, it will prolong the blacklisted status beyond easy understanding of this mechanism